// grunt-cloudfile-to-vocab/Gruntfile.js
'use strict';

module.exports = function(grunt) {

  grunt.initConfig({

    clean: {
        tests: ['test/output'],
        demo:  ['demo']
    },

    cloudfile_to_vocab: {
        task_interface: {
            options: {
                google_spreadsheet_id: '0AlnIaz1Z66BbdG8yeDBSVU1USXM4dERDQUFSZ3BTc0E',
                username:              '',
                password:              '',
                output_directory:      'demo',
                worksheets:           ['Sheet1'],
                whitelisted_services: ['arabic', 'azeri', 'bengali', 'brasil', 'burmese', 'chinese_simp', 'chinese_trad', 'cymru', 'english', 'french', 'gahuza', 'hausa', 'hindi', 'indonesian', 'kyrgyz', 'mundo', 'nepali', 'pashto', 'persian', 'portuguese', 'russian', 'sinhala', 'somali', 'swahili', 'tamil', 'turkish', 'ukrainian', 'urdu' , 'uzbek', 'vietnamese']
            }
        }
    },

    jshint: {
        all: [
            'Gruntfile.js',
            'tasks/**/*.js',
            '<%= nodeunit.tests %>',
        ],
        options: {
            "curly": true,
            "eqeqeq": true,
            "immed": true,
            "latedef": true,
            "newcap": true,
            "noarg": true,
            "sub": true,
            "undef": true,
            "boss": true,
            "eqnull": true,
            "node": true,
            "es5": true
        }
    },

    nodeunit: {
      tests: ['test/*_test.js'],
    }

  });

  grunt.loadTasks('tasks');
  grunt.loadNpmTasks('grunt-contrib-clean');
  grunt.loadNpmTasks('grunt-contrib-nodeunit');
  grunt.loadNpmTasks('grunt-contrib-jshint');

  grunt.registerTask('demo', ['clean:demo',  'cloudfile_to_vocab']);
  grunt.registerTask('test', ['jshint', 'clean:tests', 'nodeunit']);

};
